First pass of drawing API

Dinnerbone opened the pull request:
This gets it to a state where most sane applications would work, but there's still a lot left unimplemented:

- Shape bounds may contain edge cases, it's "mostly correct" but won't account for caps or joins. Considering we have no way to query shape vs edge bounds right now, I'll leave that as an exercise for another day.
- I didn't implement gradient line styles as those aren't supported in ruffle yet, so I couldn't test them.
- There's some strange logic in Flash surrounding auto-closing of shapes. I haven't fully deciphered it and so this doesn't auto-close at all.
- No bitmap stuff.

I figure it's enough to get most basic uses of the drawing API up and running, and the rest can come when they're ready.

Herschel commented on core/src/display_object/movie_clip.rs:
```diff
@@ -550,6 +578,132 @@ impl<'gc> MovieClip<'gc> {
+            if let Some(command) = mc
+                .current_fill
+                .as_ref()
+                .and_then(|(_, commands)| commands.last().cloned())
```
Don't think we need `cloned()` here, can match on `&DrawCommand::MoveTo` etc.

Dinnerbone replied:
We'd need to turn all the uses of x/y into `*x` `*y`, which I find less readable. A copy of the complete object happens one way or another, I think this way is preferable.

Herschel replied:
If I match on `&DrawCommand::MoveTo { x, y }` with the `&`, it's smart enough to match the parameters by value

Dinnerbone replied:
Added a commit that does this (but by dereferencing the match entirely, per clippy suggestions)

Herschel commented on core/src/avm1/globals/point.rs:
```diff
@@ -0,0 +1,376 @@
+    Ok(length.into())
+}
+
+fn normalize<'gc>(
```
Some ones that are currently incorrect:

`(NaN, 0).normalize()`
`(undefined, 0).normalize(1)`
`(10, null).normalize(0)`

Herschel commented on core/src/avm1/globals/point.rs:
```diff
@@ -0,0 +1,376 @@
+    args: &[Value<'gc>],
+) -> Result<ReturnValue<'gc>, Error> {
+    if args.len() < 3 {
+        return Ok(Value::Undefined.into());
```
Should return a point `(NaN, NaN)`

Herschel commented on core/src/avm1/globals/point.rs:
```diff
@@ -0,0 +1,376 @@
+    let a = value_to_point(args.get(0).unwrap().to_owned(), avm, context)?;
+    let b = value_to_point(args.get(1).unwrap().to_owned(), avm, context)?;
+    let f = args.get(2).unwrap().as_number(avm, context)?;
+    let result = (b.0 - (b.0 - a.0) * f, b.1 - (b.1 - a.1) * f);
```
Could add a small comment about the weirdo misdocument-backwards behavior

Herschel commented on core/src/avm1/globals/point.rs:
```diff
@@ -0,0 +1,376 @@
+    args: &[Value<'gc>],
+) -> Result<ReturnValue<'gc>, Error> {
+    if args.len() < 2 {
+        return Ok(Value::Undefined.into());
```
Should return `(NaN, NaN)`

Herschel commented on core/src/avm1/globals/point.rs:
```diff
@@ -0,0 +1,376 @@
+        return Ok(Value::Undefined.into());
+    }
+
+    let a = value_to_point(
+        args.get(0).unwrap_or(&Value::Undefined).to_owned(),
+        avm,
+        context,
+    )?;
+    let b = value_to_point(
+        args.get(1).unwrap_or(&Value::Undefined).to_owned(),
+        avm,
+        context,
+    )?;
```
`distance` doesn't seem to actually work for non-Point objects like `{x: 1, y: 2}` -- uses `subtract`/`length` existing on prototype

Herschel commented:
Thanks for this, it's great already

1) When a line/fill is flushed, the cursor position doesn't seem to be remembered, try:
```
lineStyle(10);
lineTo(10, 10);
lineStyle(1);
lineTo(30, 100);
lineStyle(10);
lineTo(40, 70);
```
Probably need to add an implicit MoveTo after each flush.

2) It'd be nice to move the custom shape drawing stuff out of MovieClip and into a separate module, maybe `graphics.rs`, both for organization and because it will be necessary for AS3 anyway (see https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/display/Graphics.html -- it can be used by both MovieClip and Shape in AVM2). MovieClip will own a `Graphics` (or maybe eventually an `Option<Box<Graphics>>` if we want to make it lazy initialize).

And these are more tedious nits that I don't consider blockers:

3) Matrix should not convert to/from `swf::Matrix` until actually needed somewhere, because the Matrix AS type is all `f64`s and does not use Twips, while the `swf::Matrix` type does (this causes incorrect values for large or NaN tx/ty, for example).

4) Various minor inaccuracies with weirdo Point values (see code notes).

(ugh, had a review typed up and accidentally cancelled it... so sorry if the comments are a little rushed!)

Herschel commented:
Also, with the way it's implemented currently, there can be a one-frame-off delay before the drawing appears (see the attached SWF).
[draw.zip](https://github.com/ruffle-rs/ruffle/files/4663265/draw.zip)

Dinnerbone commented:
> 
> 
> Also, with the way it's implemented currently, there can be a one-frame-off delay before the drawing appears (see the attached SWF).
> [draw.zip](https://github.com/ruffle-rs/ruffle/files/4663265/draw.zip)

That is true, unfortunately. I noticed it with the exporter. It's because render is immutable, but I can solve it with a cell until we figure out a better render API design.



>     1. When a line/fill is flushed, the cursor position doesn't seem to be remembered, try:

Fixed!



>     1. It'd be nice to move the custom shape drawing stuff out of `display_object::MovieClip` and into a separate module, maybe `graphics.rs` or `drawing.rs` or something, both for organization and because it will be necessary for AS3 anyway (see https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/display/Graphics.html -- it can be used by both MovieClip and Shape in AVM2). MovieClip will own a `Graphics` (or maybe eventually an `Option<Box<Graphics>>` if we want to make it lazy initialize).

I moved it into `drawing.rs`. I noticed there was a `Graphic` display object, but I was really unsure of its ultimate purpose. In Flash it doesn't look like there's any extra display objects for predefined or dynamic drawings, according to the debug dumps.



>     2\. Matrix should not convert to/from `swf::Matrix` until actually needed somewhere, because the Matrix AS type is all `f64`s and does not use Twips, while the `swf::Matrix` type does (this causes incorrect values for large or NaN tx/ty, for example).

That's fair, I didn't think about interior types. I did that so that we wouldn't have to duplicate existing functionality but I guess there's no way around it.



>     3\. Various minor inaccuracies with weirdo Point values (see code notes).

I may just rip Point out of this PR, it's not very topical and can wait until later!

Herschel commented on core/src/drawing.rs:
```diff
@@ -195,17 +198,26 @@ impl Drawing {
+            bounding_box.encompass(x1 - radius, y1 - radius);
+            bounding_box.encompass(x1 + radius, y1 + radius);
+            bounding_box.encompass(x2 - radius, y2 - radius);
+            bounding_box.encompass(x2 + radius, y2 + radius);
```
Better bounds for quadratic bezier: https://www.iquilezles.org/www/articles/bezierbbox/bezierbbox.htm

Herschel commented:
Thanks! It's great!

> I noticed there was a `Graphic` display object, but I was really unsure of its ultimate purpose. In Flash it doesn't look like there's any extra display objects for predefined or dynamic drawings, according to the debug dumps.

If you mean our `Graphic` display object, this actually becomes a proper display object in AS3 (all pre-drawn art is a "Shape" instance, which may be a better name for it). If you mean the `Graphics` class in AS3, it's not an actual `DisplayObject`, just a container for the shared behavior that can be used from both `Sprite` and `Shape` display objects (they aren't in the same branch of the inheritance tree, so they each have a separate instance, and a reason for us to separate it, too).
